From dc79d81ac9e9d4a9ddc49730e2f2dd699a14be22 Mon Sep 17 00:00:00 2001
From: Grzegorz Jaszczyk <jaz@semihalf.com>
Date: Thu, 19 Apr 2018 14:18:23 +0200
Subject: [PATCH 0593/1051] phy: cp110-comphy: add support for sata comphy
 initialization

The sata comphy configuration is performed by firmware, therefore the
comphy driver only triggers appropriate smc call with argument understood
by fw.

Change-Id: Ic69974ef50d6b4fdb5970bfb2aedfece9da1d2f6
Signed-off-by: Grzegorz Jaszczyk <jaz@semihalf.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/phy/marvell/phy-mvebu-cp110-comphy.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/phy/marvell/phy-mvebu-cp110-comphy.c b/drivers/phy/marvell/phy-mvebu-cp110-comphy.c
index 89414838cfdf..3f32d7500ef4 100644
--- a/drivers/phy/marvell/phy-mvebu-cp110-comphy.c
+++ b/drivers/phy/marvell/phy-mvebu-cp110-comphy.c
@@ -37,6 +37,7 @@ struct mvebu_comhy_conf {
 #define MV_SIP_COMPHY_XFI_TRAIN	0x82000004
 #define MV_SIP_COMPHY_DIG_RESET	0x82000005
 
+#define COMPHY_FW_MODE_FORMAT(mode)		((mode) << 12)
 #define COMPHY_FW_FORMAT(mode, idx, speeds)	\
 			(((mode) << 12) | ((idx) << 8) | ((speeds) << 2))
 
@@ -83,6 +84,7 @@ static const struct mvebu_comhy_conf mvebu_comphy_cp110_modes[] = {
 	MVEBU_COMPHY_CONF(1, 2, PHY_MODE_SGMII),
 	MVEBU_COMPHY_CONF(1, 2, PHY_MODE_2500SGMII),
 	MVEBU_COMPHY_CONF(1, 0, PHY_MODE_PCIE),
+	MVEBU_COMPHY_CONF(1, 0, PHY_MODE_SATA),
 	/* lane 2 */
 	MVEBU_COMPHY_CONF(2, 0, PHY_MODE_SGMII),
 	MVEBU_COMPHY_CONF(2, 0, PHY_MODE_2500SGMII),
@@ -92,6 +94,7 @@ static const struct mvebu_comhy_conf mvebu_comphy_cp110_modes[] = {
 	MVEBU_COMPHY_CONF(3, 1, PHY_MODE_SGMII),
 	MVEBU_COMPHY_CONF(3, 1, PHY_MODE_2500SGMII),
 	MVEBU_COMPHY_CONF(3, 0, PHY_MODE_PCIE),
+	MVEBU_COMPHY_CONF(3, 1, PHY_MODE_SATA),
 	/* lane 4 */
 	MVEBU_COMPHY_CONF(4, 0, PHY_MODE_SGMII),
 	MVEBU_COMPHY_CONF(4, 0, PHY_MODE_2500SGMII),
@@ -102,6 +105,7 @@ static const struct mvebu_comhy_conf mvebu_comphy_cp110_modes[] = {
 	MVEBU_COMPHY_CONF(5, 2, PHY_MODE_SGMII),
 	MVEBU_COMPHY_CONF(5, 2, PHY_MODE_2500SGMII),
 	MVEBU_COMPHY_CONF(5, 2, PHY_MODE_PCIE),
+	MVEBU_COMPHY_CONF(5, 1, PHY_MODE_SATA),
 };
 
 struct mvebu_comphy_priv {
@@ -171,6 +175,11 @@ static int mvebu_comphy_power_on(struct phy *phy)
 						       lane->port,
 						       COMPHY_SPEED_5G));
 		break;
+	case PHY_MODE_SATA:
+		ret = comphy_smc(MV_SIP_COMPHY_POWER_ON, priv->phys,
+				 lane->id,
+				 COMPHY_FW_MODE_FORMAT(COMPHY_SATA_MODE));
+		break;
 	default:
 		return -ENOTSUPP;
 	}
-- 
2.17.1

